pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,512 @@
+	}
+	klog.Infof("Workers: %d", workers)
+	for i := 0; i < workers; i++ {
+		klog.Infof("Starting worker")
+		go wait.UntilWithContext(ctx, c.runElectionWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func NewController(leaseInformer coordinationv1informers.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	c := &Controller{
+		leaseInformer:          leaseInformer,
+		leaseCandidateInformer: leaseCandidateInformer,
+		leaseClient:            leaseClient,
+		leaseCandidateClient:   leaseCandidateClient,
+
+		electionQueue: workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[leaderLeaseID](), workqueue.TypedRateLimitingQueueConfig[leaderLeaseID]{Name: controllerName}),
+	}
+	leaseSynced, err := leaseInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.processLease(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.processLease(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.processLease(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	leaseCandidateSynced, err := leaseCandidateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.processCandidate(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.processCandidate(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.processCandidate(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	c.leaseSynced = leaseSynced.HasSynced
+	c.leaseCandidateSynced = leaseCandidateSynced.HasSynced
+	return c, nil
+}
+
+func (c *Controller) runElectionWorker(ctx context.Context) {
+	klog.Infof("runElectionWorker")
+	for c.processNextElectionItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextElectionItem(ctx context.Context) bool {
+	key, shutdown := c.electionQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	completed, err := c.reconcileElectionStep(ctx, key)
+	utilruntime.HandleError(err)
+	if completed {
+		defer c.electionQueue.AddAfter(key, 5*time.Second)
+	}
+	c.electionQueue.Done(key)
+	return true
+}
+
+func (c *Controller) processCandidate(obj any) {
+	lc, ok := obj.(*v1alpha1.LeaseCandidate)
+	if !ok {
+		return
+	}
+	if lc == nil {
+		return
+	}
+	// Ignore candidates that transitioned to Pending because reelection is already in progress
+	if lc.Spec.PingTime != nil {
+		return
+	}
+
+	klog.Infof("reconcile found canLead label namespace=%q, name=%q: %q", lc.Namespace, lc.Name, lc.Spec.LeaseName)
+	leaderLeaseID := leaderLeaseID{lc.Namespace, lc.Spec.LeaseName}
+
+	klog.Infof("Adding %s to electionQueue", leaderLeaseID)
+	c.electionQueue.Add(leaderLeaseID)
+}
+
+func (c *Controller) processLease(obj any) {
+	lease, ok := obj.(*v1.Lease)
+	if !ok {
+		return
+	}
+	if lease != nil {
```

Comment:
this should never happen. Additional nil checks often have a smell of underspecification.